### PR TITLE
8282823: javac should constrain more uses of preview APIs

### PR DESCRIPTION
**Pull request by lahodaj**
Consider a class or interface with a method marked as preview. Subclassing the class or implementing the interface by itself should not cause a preview error/warning (as that would be problematic for Loom), but overriding the method should produce a preview error/warning.

This patch adds the error/warning in case a preview method is overridden.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8282823](https://bugs.openjdk.java.net/browse/JDK-8282823): javac should constrain more uses of preview APIs
 * [JDK-8283380](https://bugs.openjdk.java.net/browse/JDK-8283380): javac should constrain more uses of preview APIs (**CSR**)

### Reviewers
 * [Jim Laskey](https://openjdk.java.net/census#jlaskey) (@JimLaskey - **Reviewer**) ⚠️ Review applies to a16300adbbc36921077557e03e8960572aca837a
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to 3b8ac122710e685b541b981fed9cbacf7a6027a7
 * [Vicente Romero](https://openjdk.java.net/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to a16300adbbc36921077557e03e8960572aca837a

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7774/head:pull/7774` \
`$ git checkout pull/7774`

Update a local copy of the PR: \
`$ git checkout pull/7774` \
`$ git pull https://git.openjdk.java.net/jdk pull/7774/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7774`

View PR using the GUI difftool: \
`$ git pr show -t 7774`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7774.diff">https://git.openjdk.java.net/jdk/pull/7774.diff</a>

</details>

### CODE REVIEW

**Review by JimLaskey — APPROVED** [review]
LGTM

**Review by mcimadamore — COMMENTED** [review]
Question - what happens if you have:
```
class A {
   void m() { ... } // THIS IS PREVIEW
}

class B extends A {
    void m() { ... } // override, ok, classfile polluted
}

class C extends B {
    void m() { ... } // ?
}
```

My understanding is that C will also get a pollute classfile version - but not because of the override (because javac cannot see that the method being overriden is, in turn, the override of another preview method deep down); instead, I believe C get warning/polluted classfile because it's using B, which has a polluted classfile. But what if both B and  C are compiled together?

**Review by mcimadamore — APPROVED** [review]
If the goal is (as it seems) to detect first override of a preview method, then these changes look good.

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java [hunk 1]
```diff
@@ -1902,6 +1906,23 @@ else if (unhandledUnerased.nonEmpty()) {
         }
     }
     // where
+        private boolean shouldCheckPreview(MethodSymbol m, MethodSymbol other, ClassSymbol origin) {
+            if (m.owner != origin ||
+                //performance - only do the expensive checks when the overridden method is a Preview API:
+                (other.flags() & PREVIEW_API) == 0) {
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java:1912]: indentation seems off here - e.g. I'd expect double "tab" in the second line of the `if` condition?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java [hunk 2]
```diff
@@ -1902,6 +1906,23 @@ else if (unhandledUnerased.nonEmpty()) {
...
+                return false;
+            }
+
+            for (Symbol s : types.membersClosure(origin.type, false).getSymbolsByName(m.name)) {
+                if (m != s && m.overrides(s, origin, types, false)) {
+                    //only check if the preview only if "m" immediatelly overrides "other"
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java:1918]: This comments seems to be off

**Review by mcimadamore — APPROVED** [review]
Logic looks good. Note that `membersClosure` is memory hungry - although here you are only (correctly) only calling it when there is some preview method around, so I think that's acceptable.

**Comment by jddarcy** [comment]
Please file a CSR for the behave change under discussion. Thanks.

**Pull request closed without merging**